woocommerce/woocommerce-gutenberg-products-block
# Introduce a Cart API on top of the REST API

This PR introduces a new 'store API' under a new namespace. The goal of these new endpoints is to provide a public (no auth required) API to interact with the frontend of the store. 

Currently this includes cart, but in the future may extend to shipping, checkout and so on. This is required for blocks (being client side it needs a method of communicating with the server) and will be used by the new All Products block which needs add to cart buttons.

This API is documented inline as `@internal` use only. This will allow refinements to be made without worrying about versioning.

Closes #976

Before working on this I did test the viability of using existing AJAX endpoints, however, I quickly ran into issues due to how those endpoints work. For example, the add to cart AJAX endpoint forces "notices" to be added. This resulted in a page full of notices once navigating away from the page using the endpoint. Not ideal :) So we decides a REST API would be more appropriate.

For the schema, I opted to include some useful totals on the main `cart` endpoint which could be utilised by a client side rendered cart block in the future. Similarly for items, we are including images and prices which might be needed.

To avoid issues with nonces when logged out, and to ensure the cart exists during these requests, I needed to init the cart manually in `maybe_init_cart_session`. This did feel slightly hacky but there doesn't appear to be an alternative without modifying core.

@woocommerce/proton @claudiosanches It would be great to get your feedback on this. I think we've had quite a few requests for a cart API. Can you spot any issues with this API in it's current form? 

### Endpoints

-   GET `/wc/store/cart` - Get a representation of the cart, including totals.
-   GET `/wc/store/cart/items` - Get cart items.
-   GET `/wc/store/cart/items/<key>` - Get a single cart item.
-   PUT `/wc/store/cart/items/<key>` - Update a single cart item (quantity only).
-   POST `/wc/store/cart/items` - Create a new cart item.
-   DELETE `/wc/store/cart/items` - Delete all cart items (clear cart).
-   DELETE `/wc/store/cart/items/<key>` - Delete a single cart item.

### Screenshots

Example add to cart request and response:

![Screenshot 2019-10-15 at 16 30 31](https://user-images.githubusercontent.com/90977/66846113-3289b180-ef69-11e9-9ec8-43af4e34669a.png)

### How to test the changes in this Pull Request:

You can test this using a client such as Insomnia or Postman. If you do not use auth (and you clear cookies) you can test this as a guest. Alternatively, you can install the Basic Auth Rest API plugin and use your login credentials to test adding user carts.

Start with a POST request to `http://local.wordpress.test/wp-json/wc/store/cart/items`. Pass a valid product ID and quantity in the body, for example:

```json
{
	"quantity": 10,
	"id": 82
}
```

After this you can test the following:

1. GET `/wc/store/cart` - Check you see totals and an array of items
2. GET `/wc/store/cart/items`- Check you just see items. Note a 'key' and use this to...
3. GET `/wc/store/cart/items/<key>` - You should see a single cart item.

You can try editing an item with a PUT request to `/wc/store/cart/items/<key>`. Pass a quantity in the body:

```json
{
	"quantity": 100
}
```

See if it updates in the response.

Finally, you can clear the cart with a DELETE request to `/wc/store/cart/items`. You'll see an empty response and can confirm it worked by doing another GET `/wc/store/cart`.
